pkg/apis/flowcontrol/validation/validation.go

```diff
@@ -339,9 +341,52 @@ func ValidateFlowSchemaCondition(condition *flowcontrol.FlowSchemaCondition, fld
 	return allErrs
 }
 
+func ValidateCreatePriorityLevelConfiguration(obj *flowcontrol.PriorityLevelConfiguration, gvk schema.GroupVersion) field.ErrorList {
+	// TODO(121510): 1.29: don't allow a zero value, either via v1 or
+	//  v1beta3 (with the roundtrip annotation) for the
+	//  'nominalConcurrencyShares' field of 'limited' for CREATE operation.
+	//  1:30: lift this restriction, allow zero value via v1 or v1beta3
+	var allErrs field.ErrorList
+	if !(gvk == flowcontrolv1beta3.SchemeGroupVersion || gvk == flowcontrolv1.SchemeGroupVersion) {
+		return allErrs
+	}
+	if isNominalConcurrencyShareFieldDefault(obj) {
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec").Child("limited").Child("nominalConcurrencyShares"), "zero value is forbidden in 1.29"))
+	}
+	return allErrs
+}
+
+func ValidateUpdatePriorityLevelConfiguration(obj, old *flowcontrol.PriorityLevelConfiguration, gvk schema.GroupVersion) field.ErrorList {
+	// TODO(121510): 1.29: only allow a zero value, either via v1 or
+	//  v1beta3 (with the roundtrip annotation) for the
+	//  'nominalConcurrencyShares' field of 'limited' for UPDATE operation,
+	//  only if the existing object already contains a zero value.
+	//  1:30: lift this restriction, allow zero value via v1 or v1beta3
+	var allErrs field.ErrorList
+	if !(gvk == flowcontrolv1beta3.SchemeGroupVersion || gvk == flowcontrolv1.SchemeGroupVersion) {
+		return allErrs
+	}
+	if !isNominalConcurrencyShareFieldDefault(obj) {
+		return allErrs
+	}
+
+	// the existing object must have a zero value as well
+	if !isNominalConcurrencyShareFieldDefault(old) {
+		fd := field.NewPath("spec").Child("limited").Child("nominalConcurrencyShares")
+		allErrs = append(allErrs, field.Forbidden(fd, "setting a zero value is forbidden in 1.29, unless the existing object has a zero value as well"))
+	}
+	return allErrs
+}
+
 // ValidatePriorityLevelConfiguration validates priority-level-configuration.
 func ValidatePriorityLevelConfiguration(pl *flowcontrol.PriorityLevelConfiguration, requestGV schema.GroupVersion) field.ErrorList {
 	allErrs := apivalidation.ValidateObjectMeta(&pl.ObjectMeta, false, ValidatePriorityLevelConfigurationName, field.NewPath("metadata"))
+
+	// the roundtrip annotation is forbidden in v1
```

Comment:
the annotation is forbidden in the internal version no matter how we got here... it gets dropped in conversion from v1beta3 → internal so we should never see it here

Reply:
I am thinking of the following scenario in 1.30:
- client (using v1beta3) sets the annotation to "true" and nominalConcurrencyShares to zero
- during conversion to internal, we drop the annotation
- when we store to:
      --  v1beta3: the annotation gets added to ""
      --  v1: no annotation 
- when we serve,  the client sees a value of "" which is different from the originally specified value.

the client is expected to work with the object returned from the create/update response, so i think it should't be a problem, but if you think this could pose other problems, may be we should enforce the annotation to be an empty string?